Initial Integration of DAPI

NiteKat opened the pull request:
This commit adds the DAPI server, the ProtoClient that the server uses to communicate with a front end client, with the necessary CMake file changes to integrate the new files.

Some code was moved from anonymous namespaces in DevilutionX to the devilution namespace so that DAPI server can access them.

NiteKat committed:
Initial Integration of DAPI

This commit adds the DAPI server,the ProtoClient that the server uses to communicate with a front end client, and the supporting SFML networking files, with the necessary CMake file changes to integrate the new files.

Some code was moved from anonymous namespaces in DevilutionX to the devilution namespace so that DAPI server can access them.

This builds, but crashes at the SFML initialization.

AJenbo committed:
Fix building on linux and use unchanged SFML

NiteKat committed:
SFML and ProtoClient Fixes

Change to newest SFML and fix linkage, additionally fix the DAPIProtoClient to be SFML 3.0 compatible.

AJenbo committed:
Fix warnings

NiteKat committed:
Fix Monster Selection Loop and DAPI Griswold

Loop crashes (in debug build only somehow?) for selecting monster types when playing Diablo mode due to less monster types being available.

Fixes Griswold data not being sent client side due to extra blank entries in Towners struct.

NiteKat committed:
Fix Inventory Functions

Fixes compatibility for the inventory manipulation functions for DevilutionX.

AJenbo committed:
Merge branch 'master' into dapiintegration

NiteKat committed:
FIx Menu Actions

This allows DAPI to interact with the menu in game.

AJenbo commented on Source/dapi/Server.cpp:
```diff
@@ -2401,6 +2401,7 @@ void Server::saveGame()
 
+	devilution::gmenu_presskeys(SDLK_DOWN);
 	devilution::gmenu_presskeys(SDLK_KP_ENTER);
 	return;
```
DevilutionX has quick save:
```cpp
if (devilution::gbIsMultiplayer)
		return;
gamemenu_save_game(false);
```

NiteKat replied:
Does it make the menu sounds though? I could just add a menu sound in like I did for 1.09 though. 😆

AJenbo replied:
The human player can do this, it's bound to one of the F-keys by default.

It shows the notification, but no it doesn't play the sound.

NiteKat committed:
Translate interface_mode

Frontend uses the custom defined windows message values that 1.09 uses internally. This translates the equivalent DevilutionX messages back to the expected values.

NiteKat committed:
Fix Command Send Changes

Fixes CMD_OPOBJXY and CMD_DISARMXY which are sent with a different NetSend command in DevilutionX.

NiteKat committed:
Refactor and Fix Inv Paste and Witch Sell

Fixes Inventory pasting and refactors it based on how DevilutionX works internally. Also fixes Adria's sell window storing wrong OldActiveStore.

NiteKat committed:
Add Backend Awareness

Adds an enum and to the frameupdate message so the frontend can know what backend it is attached to. Needed since DevilutionX has some differences such as TileIDs and Pepin auto healing, etc.

NiteKat committed:
Fix Placing Items From Cursor

This makes placing items from cursor functional as it fully accounts for how DevilutionX handles item graphic cursors (the cursor point is considered in the middle instead of top left)

StephenCWills committed:
Enable compiling SFML from source

NiteKat committed:
Prevent Improper Store Usage

Store options are built even if quest test is being displayed, so this prevents a DAPI AI from using a store while a quest text is up.

NiteKat committed:
Merge pull request #1 from StephenCWills/dapi-sfml-from-source

Enable compiling SFML from source

NiteKat committed:
Fix Towners

Fix towners to that cows don't get squished into one entity.

NiteKat committed:
Monster Name and Witch Fix

Fixes the Monster Name passed to client and fixes buying items from Adria.

NiteKat committed:
Fix Quest Text

Fix quest text by adding a space between the line breaks.

NiteKat committed:
Fix Repair Logic

Fix issue where magical items can have repair cost under 1 gold, and thus not be eligible for repair.

NiteKat committed:
Merge branch 'master' into dapiintegration

NiteKat committed:
Fix putCursorItem

Need to explicitly check if HoldItem is empty in DevX

NiteKat committed:
Fix Repairs and Identifying with Cain

This actually fixes repairs (and selling), using CurrentItemIndex correctly and doesn't double dip into the players gold for repairs. Also fixes Cain identifies as before the store wasn't being re-ran after identifying an item.

NiteKat committed:
Fix Player Information

Fixes issue where player information was being given when players were not on the active floor.

NiteKat committed:
Merge branch 'master' into dapiintegration

NiteKat committed:
Fix Player Info Loop and Implement FPS Change

NiteKat committed:
Merge branch 'master' into dapiintegration

StephenCWills committed:
Merge branch 'master' into dapiintegration

StephenCWills committed:
Fix path separator in DAPI Server.cpp includes

StephenCWills committed:
Add CMake option to opt-in to DAPI

NiteKat committed:
Merge pull request #2 from StephenCWills/dapi-option

Add CMake option to opt-in to DAPI

NiteKat committed:
Merge branch 'dapiintegration' of https://github.com/NiteKat/devilutionX into dapiintegration

NiteKat committed:
Add Initial Chat Functionality

Moves three chat log items to devilution namespace and adds commands for reading from and writing to chat.

StephenCWills committed:
Apply clang-format

StephenCWills commented:
Hrm.. maybe the clang-format check shouldn't be looking at the protobuf files?

NiteKat commented on CMakeSettings.json:
```diff
@@ -12,7 +12,7 @@
         {
           "name": "DISCORD_INTEGRATION",
-          "value": "False",
+          "value": "True",
```
Which commit is that from? When I build I don't have the patch.exe or whatever is needed so I typically don't build with discord integration. I leave out that change in the CMakeSettings.json when I push, but maybe it slipped in on accident on a push?

NiteKat replied:
Thanks! I think it makes sense to have that integration disabled on the DAPI configuration, seems like you wouldn't really need (want?) the discord features for the AI clients.

glebm commented:
clang-format does handle proto files but we should use the default Google configuration for those as that's what everybody else uses

NiteKat commented:
> > Probably not since those are generated, though in the long run they may not be included in PR.
> 
> I'm not sure I understand. The C++ files are generated from the proto files, but are the proto files generated from something else? Also, are you saying they may not be included in the PR because you're considering deduplicating the data based on your discussion with rouming?
> 
> FYI, the `clang-format` CI doesn't actually build the project so it's checking the syntax in the proto files, not the generated files. My reasoning is that the proto files technically aren't C-lang. On the other hand, it's close enough in syntax that it should be possible to format the proto files so we might want to format them after all, to enforce style rules. 🤷

Oh sorry, I thought you meant the C++ files that get generated from the .proto files, but it looks like I'm confused as I don't see those files in the project so I think they got removed already, or generally aren't committed? I'll look into it a bit more, but I thought you were saying to not apply the clang-format check to those generated files, not the .proto files.

NiteKat committed:
Chat Message Changes

This will need to be changed again to fix longer chat messages, but as long as a message doesn't trigger wordwrap in the message log this works.

NiteKat committed:
Add Map Include to Player.h

Adds Map Include to Player.h to remove clang-tidy warnings

NiteKat committed:
Merge branch 'master' into dapiintegration

NiteKat committed:
Merge branch 'master' into dapiintegration

NiteKat committed:
Fix Handling Item Purchases

Fixes handling of item purchases to work with the new StaticVector class and corrects a DevilutionX function definition that for some reason didn't update when I merged master.

NiteKat committed:
Add _pISplLvlAdd to PlayerData

Adds collecting _pISplLvlAdd to PlayerData server side and sending it client side.

NiteKat committed:
Fix API Item Dropping

Fixes an issue that started after updating to newer commit. Moves TryOpenDungeonWithMouse() to devilution namespace for access in the API and copies part of the left click code with slight modification to handle dropping items, and sets me up for implementing API opening hellfire areas.

NiteKat committed:
Add Stash Support Part 1

Adds the start of support for the API to utilize the stash. Gold deposits and withdrawl is handled, and technically the API can place and remove items from the stash; however I believe I still need to add page controls and providing the layout of the active page through the API.

NiteKat committed:
Fix Issue With Item Placement In Inventory

Fixes an issue with hotPixelCellOffset.

NiteKat committed:
Catching dapiintegration up with master and resolving merge conflicts

NiteKat committed:
Fix Up Stores

Fix some errors introduced resolving merge conflicts - accidentally added extra namespace and some duplicate definitions. Visual store feature pulled some things out of anonymous namespace that I already pulled out.

NiteKat committed:
Fix Compile and Linker Errors

spelldat.h include removed from Server.h, and restored a definition removed from stores.cpp on accident.

NiteKat committed:
Add x64-Release-DAPI Configuration

Adds a release build configuration for DeviltuionX with DAPI.

NiteKat committed:
Merge branch 'diasurgical:master' into dapiintegration
